### PR TITLE
proposal: Add proposal for a chrooted kubelet

### PR DESCRIPTION
**Pull request by euank**
After discussion with @philips and @vishh we settled on running the Kubelet in a chroot as being a fairly sane way to solve the problem of mount utility availability (among a couple others).

This proposal outlines a few more details of that idea and provides some alternatives.

Related issues: #19765, #16508, #35224, #35249 and others I'm sure.

cc @vishh @philips

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35328)

### CODE REVIEW

**Comment by luxas** [comment]
cc @kubernetes/sig-cluster-lifecycle this affects our packaging work as well

**Comment by euank** [comment]
cc @kubernetes/sig-node, we discussed this a bit in our meeting today.

If there was a group to ping for gci-related-things I'd include that here, but I don't know of one.

File: docs/proposals/kubelet-rootfs-distribution.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+The Kubelet is a critical component of Kubernetes that must be run on every node in a cluster.
+
+However, right now it's not always easy to run it *correctly*. The Kubelet has
+a number of dependencies that must exist in its filesystem, including various
+mount and network utilities. Missing any of these can lead to unexpected
+differences between Kubernetes hosts. For example, the Google Container VM
+image (GCI) is missing various mount commands even though the Kernel supports
+those filesystem types. Similarly, CoreOS Linux intentionally doesn't ship with
+many mount utilities or socat in the base image.
```
Comment (vishh) [inline docs/proposals/kubelet-rootfs-distribution.md]: Other distros can also benefit from this proposal. For instance, requiring users to manually install a bunch of packages is tedious. Also these packages aren't version controlled if they are installed out of band. Shipping kubelet in a chroot sandbox along with all of its tested dependencies will benefit all users IMHO

File: docs/proposals/kubelet-rootfs-distribution.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+The easiest way to express all filesystem dependencies of the Kubelet comprehensively is to ship a filesystem image and run the Kubelet within it. The [hyperkube image](../../cluster/images/hyperkube/) already provides such a filesystem.
+
+Even though the hyperkube image is distributed as a container, this method of
+running the Kubelet intentionally is using a chroot and is neither a container nor pod.
+
+The kubelet chroot will essentially operate as follows:
+
+```
+container-download-and-extract gcr.io/google_containers/hyperkube:v1.4.0 /path/to/kubelet
+mount --make-shared /var/lib/kubelet
```
Comment (vishh) [inline docs/proposals/kubelet-rootfs-distribution.md]: For hostpath volumes to function, the entire rootfs `/` has to be mounted in. So its better to expose `/`  under `/media/root` or `/mnt/root` inside the chroot.

In addition to that, kubelet volumes plugins have to take an additional & optional `rootfs` path which will ensure that they access directories on the real rootfs on the host.

File: docs/proposals/kubelet-rootfs-distribution.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+mount --rbind /var/lib/kubelet /path/to/kubelet/var/lib/kubelet
+mount --rbind /sys /path/to/kubelet/sys
```
Comment (vishh) [inline docs/proposals/kubelet-rootfs-distribution.md]: If you intend to make this complete, include `procfs`, `/dev` and `cgroupfs` at the least here.

File: docs/proposals/kubelet-rootfs-distribution.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+#### Hyperkube Image Packaging
+
+The Hyperkube image is distributed as part of an official release to the `gcr.io/google_containers` registry, but is not included along with the `kube-up` artifacts used for deployment.
+
+This will need to be remediated in order to complete this proposal.
+
+#### Everything else
+
+In the initial implementation, rkt or docker can be used to extract the rootfs of the hyperkube image. rkt fly or a systemd unit (using `RootDirectory`) can be used to perform the needed setup, chroot, and execution of the kubelet within that rootfs.
```
Comment (vishh) [inline docs/proposals/kubelet-rootfs-distribution.md]: How can systemd be used without rkt?

File: docs/proposals/kubelet-rootfs-distribution.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
```
Comment (vishh) [inline docs/proposals/kubelet-rootfs-distribution.md]: Add a testing and rollout plan. The way I see it, we will have to identify & clone a subset of high value e2e tests to run kubelet in a chroot. Once we gain confidence, we need to enable this across all e2es. 
Given that there is only one feature change required for this feature, we should target v1.6.

File: docs/proposals/kubelet-rootfs-distribution.md (hunk 4 above)
Comment (philips) [inline docs/proposals/kubelet-rootfs-distribution.md]: Probably should just link to this: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RootDirectory=

File: docs/proposals/kubelet-rootfs-distribution.md (hunk 2 above)
Comment (vishh) [inline docs/proposals/kubelet-rootfs-distribution.md]: As of today, users can mount in arbitrary pre-existing directories or files on the host. So irrespective of #34058, we will need to support hostpath volumes.

File: docs/proposals/kubelet-rootfs-distribution.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+The kubelet chroot will essentially operate as follows:
+
+```
+container-download-and-extract gcr.io/google_containers/hyperkube:v1.4.0 /path/to/kubelet
+mount --make-shared /var/lib/kubelet
+mount --rbind /var/lib/kubelet /path/to/kubelet/var/lib/kubelet
+mount --rbind /sys /path/to/kubelet/sys
+...
+chroot /path/to/kubelet /usr/bin/hyperkube kubelet
```
Comment (fabiand) [inline docs/proposals/kubelet-rootfs-distribution.md]: The steps from line 60-65 -- Isn't that something which could be put into a systemd portable service?

File: docs/proposals/kubelet-rootfs-distribution.md (hunk 3 above)
Comment (fabiand) [inline docs/proposals/kubelet-rootfs-distribution.md]: The issue I see with explicitly listing the host sided paths to pull into the kubelets chroot is, that in future additional paths might be required - in that case this list needs to be updated.

**Review by fabiand — COMMENTED** [review]
Just a few inline comments.

**Comment by philips** [comment]
@euank @vishh I think this needs to move here now: https://github.com/kubernetes/community/tree/master/contributors/design-proposals

**Comment by euank** [comment]
thanks for the heads up.
I'll move it over tomorrow and address comments as well. Sorry for letting this slip off the radar a bit!

**Comment by euank** [comment]
Moved to https://github.com/kubernetes/community/issues/131

**Pull request closed without merging**